addons/mail/models/mail_channel.py

```diff
@@ -90,6 +105,113 @@ def default_get(self, fields):
+    auto_notification = fields.Boolean(string="Automatic notification", help="People receive an automatic notification about their message being waiting for moderation.")
+    notification_message = fields.Text(string="Notification message")
+    send_guidelines = fields.Boolean(string="Send guidelines to new subscribers", help="Newcomers on this moderated channel will automatically receive the guidelines.", default=False)
+    guidelines = fields.Text(string="Guidelines")
+
+    @api.constrains('moderator_ids')
+    def _check_moderators_have_email_addresses(self):
+        if self.mapped('moderator_ids').filtered(lambda moderator: moderator.email == False):
+            raise ValidationError("Moderators must have an email address!")
+
+
+    @api.constrains('moderator_ids', 'channel_partner_ids', 'channel_last_seen_partner_ids')
+    def _check_moderators_are_channel_partners(self):
+        for channel in self:
+            if not (channel.mapped('moderator_ids.partner_id') <= channel.channel_partner_ids):
+                raise ValidationError("Moderators cannot unsubscribe from a channel they moderate!")
+
+    @api.constrains('moderation', 'email_send')
+    def _check_moderation_implies_email_send(self):
+        if self.filtered(lambda channel: channel.moderation and not channel.email_send):
+            raise ValidationError('Only email lists can be moderated!')
+
+    @api.onchange('moderator_ids')
+    def _add_moderators_to_channel_last_seen_partners(self):
+        ChannelPartner = self.env['mail.channel.partner']
+        for partner in (
+            self.mapped('moderator_ids.partner_id')
+            - self.mapped('channel_last_seen_partner_ids.partner_id')
+        ):
+            channel_partner = ChannelPartner.new({'partner_id': partner.id})
+            self.channel_last_seen_partner_ids += channel_partner
+
+    @api.onchange('email_send')
+    def _set_moderation_to_false_if_email_send_false(self):
+        if not self.email_send:
+            self.moderation = False
+
+    @api.onchange('moderation')
+    def _clear_some_data_and_accept_messages_if_moderation_set_to_false(self):
+        if not self.moderation:
+            self.auto_notification = False
+            self.send_guidelines = False
+            self.moderator_ids = self.env['res.users']
+            if self._origin:
+                MessagesToAccept = self.env['mail.message'].search(
+                        [
+                            ('moderation_status', '=', 'pending_moderation'),
+                            ('model', '=', 'mail.channel'),
+                            ('res_id', '=', self._origin.id)
+                        ]
+                    )
+                MessagesToAccept.accept_message()
+
+    @api.multi
+    def send_guidelines_to_partners(self):
+        """ Send guidelines to all channel followers.
+            Security problems?  
```

Comment:
What you can do to avoid security issues and/or information leak is to check if the current user is either admin, either one of the channel moderator.